File: 백준/Gold/9663. N－Queen/N－Queen.swift
var input = Int(readLine()!)!
var board = Array(repeating: Array(repeating: 0, count: input), count: input)
var visitedColumn = Array(repeating: false, count: input)
var count = 0

func DFS(depth: Int, column: Int) {

    if depth == (input-1) {
        count += 1
        return
    }
    // depth = 1 이고 두번째 줄 채웠고, 3번째 보고있다!.
    // k가 다음에 올 column이다!..
    for k in 0..<input {
        var isOk = true
        if visitedColumn[k] {
            continue
        }
        for m in 0...depth {
            // 한 뎁스끼리 가면서 대각선 확인하기!..
            if k-m-1 >= 0 {
                if board[depth-m][k-m-1] == 1 {
                    isOk = false
                    break
                }
            }

            if k+m+1 < input {
                if board[depth-m][k+m+1] == 1 {
                    isOk = false
                    break
                }
            }
        }
        // 돌렸는데 만약 대각선에 1이 존재했으면!.. continue 돌리기!~
        if !isOk {
            continue
        }

        visitedColumn[k] = true
        board[depth+1][k] = 1
        DFS(depth: depth+1, column: k)
        visitedColumn[k] = false
        board[depth+1][k] = 0
    }

}

for i in 0..<input {
    visitedColumn[i] = true
    board[0][i] = 1
    DFS(depth: 0, column: i)
    visitedColumn[i] = false
    board[0][i] = 0
}

print(count)

File: 백준/Gold/9663. N－Queen/N－Queen.py
N = int(input())

ans = 0
row = [0] * N
def is_promising(x):
    for i in range(x):
        if row[x] == row[i] or abs(row[x] - row[i]) == abs(x - i):
            return False

    return True


def n_queens(x): # x가 depth이자 row라고 생각하면 될듯!..
    global ans
    if x == N:
        ans += 1
        return

    else:
        for i in range(N): # 하나의 row에 대해서 각각의 column을 순회해서 확인해본다.
            # [x, i]에 퀸을 놓겠다. -> row - x, column - i 로 설정 
            row[x] = i
            if is_promising(x):
                n_queens(x + 1)


n_queens(0)
print(ans)